_posts/2020-09-07-sparklyr-flint/sparklyr_flint.Rmd

```diff
@@ -0,0 +1,227 @@
+
+sc <- spark_connect(master = "yarn-client", spark_home = "/usr/lib/spark")
+```
+
+### Importing external data to Spark
+
+Making external data available in Spark is easy with `sparklyr` given the large number of data sources `sparklyr` supports. For example, given a R dataframe, such as
+
+```{r, eval = FALSE, echo = TRUE, indent = "  "}
+dat <- data.frame(id = seq(10), value = rnorm(10))
+```
+,
+
+the command to copy it to a Spark dataframe with 3 partitions is simply
+
+```{r, eval = FALSE, echo = TRUE, indent = "  "}
+sdf <- copy_to(sc, dat, name = "unique_name_of_my_spark_dataframe", repartition = 3L)
+```
+.
+
+Similarly, there are options for ingesting data in CSV, JSON, ORC, AVRO, and many other well-known formats into Spark as well:
+
+```{r, eval = FALSE, echo = TRUE, indent = "  "}
+sdf_csv <- spark_read_csv(sc, name = "another_spark_dataframe", path = "file:///tmp/file.csv", repartition = 3L)
+# or
+sdf_json <- spark_read_json(sc, name = "yet_another_one", path = "file:///tmp/file.json", repartition = 3L)
+# or spark_read_orc, spark_read_avro, etc
+```
+.
+
+### Transforming a Spark dataframe
+
+With `sparklyr`, the simplest and most readable way to transformation a Spark dataframe is by using `dplyr` verbs and the pipe operator (`%>%`) from [magrittr](https://cran.r-project.org/web/packages/magrittr/index.html).
+
+`Sparklyr` supports a large number `dplyr` verbs. For example,
+
+```{r, eval = FALSE, echo = TRUE, indent = "  "}
+sdf <- sdf %>%
+  dplyr::filter(!is.null(id)) %>%
+  dplyr::mutate(value = value ^ 2)
+```
+
+ensures `sdf` only contains rows with non-null IDs, and then squares the `value` column of each row.
+
+# What is `Flint`?
+
+`Flint` is a powerful open-source library for working with time-series data in Apache Spark. First of all, it supports efficient computation of aggregate statistics on time-series data points having the same timestamp (a.k.a `summarizeCycles` in `Flint` nomenclature), within a given time window (a.k.a., `summarizeWindows`), or within some given time intervals (a.k.a `summarizeIntervals`). It can also join two or more time-series datasets based on inexact match of timestamps using asof join functions such as `LeftJoin` and `FutureLeftJoin`. The author of `Flint` has outlined many more of `Flint`'s major functionalities in [this article](https://databricks.com/blog/2018/09/11/introducing-flint-a-time-series-library-for-apache-spark.html), which I found to be extremely helpful when working out how to build `sparklyr.flint` as a simple and straightforward R interface for such functionalities.
+
+Readers wanting some direct hands-on experience with Flint and Apache Spark can go through the following steps to run a minimal example of using Flint to analyze time-series data:
+
+- First, install Apache Spark locally, and then for convenience reasons, define the `SPARK_HOME` environment variable. In this example, we will run Flint with Apache Spark 2.4.4 installed at `~/spark`, so:
```

Comment:
In future posts, we probably want to avoid showing Scala code, mostly since it's confusing to R users and we probably also don't want to introduce `sparklyr` again. I think it's fine for this one, serves as a refresher and gives some perspective on Scala vs R, but probably worth avoiding in subsequent posts.